cw/capture.py

```diff
@@ -1167,118 +1169,269 @@ def capture_otbn_vertical(ot, ktp, fw_bin, pll_frequency, capture_cfg, device_cf
+            # expectations.
+            share0 = ot.target.simpleserial_read("r", seed_bytes, ack=False)
+            share1 = ot.target.simpleserial_read("r", seed_bytes, ack=False)
+            if share0 is None:
+                raise RuntimeError('Random share0 is none')
+            if share1 is None:
+                raise RuntimeError('Random share1 is none')
+
+            d0 = int.from_bytes(share0, byteorder='little')
+            d1 = int.from_bytes(share1, byteorder='little')
+            actual_key = (d0 + d1) % curve_order_n
+
+            tqdm.write(f'share0 = {share0.hex()}')
+            tqdm.write(f'share1 = {share1.hex()}')
+
+            if actual_key != expected_key:
+                raise RuntimeError('Bad generated key:\n'
+                                   f'Expected: {hex(expected_key)}\n'
+                                   f'Actual:   {hex(actual_key)}')
+
+            # Create a chipwhisperer trace object and save it to the project
+            # Args/fields of Trace object: waves, textin, textout, key
+            textout = share0 + share1  # concatenate bytearrays
+            trace = Trace(waves, mask, textout, seed_used)
+            check_range(waves, ot.scope.adc.bits_per_sample)
+            project.traces.append(trace, dtype=np.uint16)
+    elif capture_cfg["app"] == 'modinv':
+        # Check the lengths in the key/plaintext generator. In this case, "key"
+        # is the input to the modinv app. We only use the key part of the ktp
+        # to generate the key share inputs to the modinv app.
+        if ktp.keyLen() != modinv_share_bytes:
+            raise ValueError(
+                f'Unexpected input (share) length: {ktp.keyLen()}.\n'
+                f'Hint: set key len={modinv_share_bytes} in the configuration file.'
+            )
+
+        # select the otbn app on the device (0 -> keygen, 1 -> modinv)
+        ot.target.simpleserial_write("a", bytearray([0x01]))
+        time.sleep(0.3)
+
+        # set PRNG seed to ensure same sequence of randoms for split up measurements
+        random.seed(capture_cfg["batch_prng_seed"])
+
+        # set ecc256 fixed key and share inputs
+        # (uncomment the desired fixed shares depending on whether
+        # you want a random fixed key or a hardcoded fixed key)
+        # k_fixed_barray = ktp.next_key()
+        k_fixed_barray = bytearray(
+            (0x2648d0d248b70944dfd84c2f85ea5793729112e7cafa50abdf7ef8b7594fa2a1
+             ).to_bytes(key_bytes, 'little'))
+        k_fixed = int.from_bytes(k_fixed_barray, byteorder='little')
+        # generate two random shares
+        input_k0_fixed = ktp.next_key()
+        input_k1_fixed = ktp.next_key()
+        k0_fixed = int.from_bytes(input_k0_fixed, byteorder='little')
+        k1_fixed = int.from_bytes(input_k1_fixed, byteorder='little')
+        # adapt share k1 so that k = (k0 + k1) mod n
+        k_tmp = (k0_fixed + k1_fixed) % curve_order_n
+        k_tmp_diff = k_fixed - k_tmp
```

Comment:
Hmm I don't think this is necessary, because a negative difference, in my opinion, is not problematic. Actually, when k_tmp is bigger than k_fixed we need to subtract the difference from one of the shares to get k_fixed as a result from the modulo operation of the share sum. If the difference would be positive, even if k_tmp is bigger than k_fixed, in my opinion this would not lead to k_fixed as result from the modulo operation of the share sum.

Reply:
But if the k_tmp_diff is negative then there is a slight chance k1_fixed may become negative, right?

Reply:
Ahh yes true, didn't have in mind, that theoretically k1_fixed can be for example 320-bit zero.

Reply:
Exactly!